docs/dev/relay_bring_your_own_codegen.rst

```diff
@@ -0,0 +1,498 @@
+
+Code Generation for Input Variables
+-----------------------------------
+
+Recall that we collected the input buffer information by visiting the arguments of a call node (2nd step in the previous section), and handled the case when its argument is another call node (4th step). In this section, we demonstrate how to handle other nodes by taking ``VarNode`` as an example.
+
+``VarNode`` represents input tensors in a model. The only but important information it has is a name hint (e.g., ``data``, ``weight``, etc). When visiting a ``VarNode``, we simply update class variable ``out_`` to pass the name hint so that the descendant call nodes can generate the correct function call.
+
+.. code-block:: c++
+
+  void VisitExpr_(const VarNode* node) {
+    ext_func_args_.push_back(node->name_hint());
+    out_.clear();
+    out_.push_back({node->name_hint(), 0});
+  }
+
+Note that in this example we assume the subgraph we are offloading has only call nodes and variable nodes. If your subgraphs contain other types of nodes, such as ``TupleNode``, then you also need to visit them and bypass the output buffer information.
+
+Code Emitting
+-------------
+
+The final part in this codegen class is a ``JIT`` function that emits the compilable C code. Remember, in addition to the subgraph function we generated in the previous sections, we also need a wrapper function (``json_rt_0``) with a unified argument for TVM runtime to invoke and pass data. Fortunately, the base class we inherited already provides an implementation, ``JitImpl``, to generate such functions. The only thing we need to do is passing all subgraph function code we generated to ``JitImpl``:
+
+.. code-block:: c++
+
+  std::string JIT() {
+    // Write function macros
+    for (auto decl : func_decl_) {
+      code_stream_ << decl << "\n";
+    }
+    return JitImpl(ext_func_id_, ext_func_args_, buf_decl_, ext_func_body, out_);
+  }
+
+All variables (``ext_func_id``, etc) we passed are class variables and were filled when we traversed the subgraph.
+
+Implement CSourceCodegen
+========================
+
+Again, let's create a class skeleton and implement required functions. Note that it inherits ``CSourceModuleCodegenBase``
+
+.. code-block:: c++
+
+  class CSourceCodegen : public CSourceModuleCodegenBase {
+   public:
+    // Pass a subgraph function, and generate the C code.
+    void GenCFunc(const Function& func) { ; }
+
+    // Use GetFunc to generate the C code and wrap it as a C source module.
+    runtime::Module CreateCSourceModule(const NodeRef& ref) override { ; }
+
+   private:
+    std::ostringstream code_stream_;
+  };
+
+Implement GetFunc
+-----------------
+
+``GetFunc`` simply uses the ``CodegenC`` we just implemented to traverse a Relay function (subgraph) and obtains the generated C code. The builtin function ``GetExtSymbol`` retrieves a unique symbol name in the Relay function and we could use it as the C function name directly.
```

Comment:
We way want to mention that this symbol is also going to be used for DSO runtime lookup